generate_report.py

```diff
@@ -0,0 +1,532 @@
+"""HummingBird Report methods.
+
+The file contain methods related to output report.
+
+"""
+import datetime
+import tempfile
+temp_path = tempfile.gettempdir()
+
+__main__ = ["SVGfile", "OutputReportFile"]
+
+
+def SVGfile(data, data_max, data_min, rect_idx, rect_width, field):
+  """Generate SVG plot.
+
+  Args:
+    data:  numpy array of voltages
+    data_max:  data maximum value
+    data_min:  data minimum value
+    rect_idx:  index where the worst pattern occurs
+    rect_width:  the index width where the worst pattern last
+    field:  the SPEC field of this SVF plot
+
+  Returns:
+    SVG plot to write in the html report.
+  """
+  svgfile = ""
+  data_max = int(data_max*50)
+  data_min = int(data_min*50)
+  height = (data_max - data_min)*2
+  rate = min(max(len(data) // 1000, 1), 180)
+  width = len(data) // rate * 5
+
+  if field == "scl_show":
+    svgfile += (
+        f"<div id='{field}'><div class='column_left'>SCL capture</div>"
+        "<div class='column_right'>"
+        f"<svg viewBox='0 0 {width} {height*1.2}' height={height*0.7} width=auto>"
+    )
+
+  elif field == "sda_show":
+    svgfile += (
+        f"<div id='{field}'><div class='column_left'>SDA capture</div>"
+        "<div class='column_right'>"
+        f"<svg viewBox='0 0 {width} {height*1.2}' height={height*0.7} width=auto>"
+    )
+
+  else:
+    if "sda" in field:
+      svgfile += (
+          f"<div id='{field}_hide' class='hide'>"
+          "<div class='column_left'>SDA capture</div><div class='column_right'>"
+      )
+
+    else:
+      svgfile += (
+          f"<div id='{field}_hide' class='hide'>"
+          "<div class='column_left'>SCL capture</div><div class='column_right'>"
+      )
+    svgfile += (
+        f"<svg viewBox='0 0 {width} {height*1.2}' height={height*0.7} width=auto>"
+    )
+
+    # Red Rect to Mark the Measure Area
+
+    rect_width = max(rect_width//rate*5, 7)
+    rect_x = rect_idx//rate*5 - rect_width
+    svgfile += (
+        f"<rect x={rect_x} y='0' width={rect_width} height='100%' fill='red' opacity='0.3'/>"
+    )
+
+  # Data Polyline
+
+  points = ""
+  for i in range(0, len(data), rate):
+    points += f"{i//rate*5},{(data_max - int(data[i]*50))*2 + 40} "
```

Comment:
I guess the string concat is slow when data is very large. You can try to output to the html file directly